compiler/backend_inkwell/src/lib.rs

```diff
@@ -0,0 +1,1041 @@
+                        self.locals.insert(*id, call_value.clone());
+
+                        if idx == mir.expressions.len() - 1 {
+                            self.builder
+                                .build_return(Some(&call_value.into_pointer_value()));
+                        }
+                    }
+                }
+                candy_frontend::mir::Expression::UseModule { .. } => unreachable!(),
+                candy_frontend::mir::Expression::Panic { reason, .. } => {
+                    let panic_fn = self.module.get_function("candy_panic").unwrap();
+
+                    let reason = self.get_value_with_id(function_ctx, reason).unwrap();
+
+                    self.builder.build_call(panic_fn, &[reason.into()], "");
+
+                    self.builder.build_unreachable();
+                }
+                candy_frontend::mir::Expression::TraceCallStarts { .. } => unimplemented!(),
+                candy_frontend::mir::Expression::TraceCallEnds { .. } => unimplemented!(),
+                candy_frontend::mir::Expression::TraceExpressionEvaluated { .. } => {
+                    unimplemented!()
+                }
+                candy_frontend::mir::Expression::TraceFoundFuzzableFunction { .. } => {
+                    unimplemented!()
+                }
+            }
+        }
+    }
+
+    fn make_str_literal(&self, s: &str) -> ArrayValue<'_> {
+        let i8_type = self.context.i8_type();
+        let content: Vec<_> = s
+            .chars()
+            .chain(std::iter::once('\0'))
+            .map(|c| i8_type.const_int(c as u64, false))
+            .collect();
+        i8_type.const_array(&content)
+    }
+
+    fn get_value_with_id(
+        &self,
+        function_ctx: &FunctionInfo<'ctx>,
+        id: &Id,
+    ) -> Option<BasicValueEnum<'ctx>> {
+        let candy_type_ptr = self
+            .module
+            .get_struct_type("candy_type")
+            .unwrap()
+            .ptr_type(AddressSpace::default());
+        let mut v = self.values.get(id).map(|a| {
+            self.builder
+                .build_load(candy_type_ptr, a.as_pointer_value(), "")
+        });
+        if v.is_none() {
+            if let Some(i) = function_ctx.captured_ids.iter().position(|i| i == id) {
+                let env_ptr = function_ctx.function_value.get_last_param().unwrap();
+
+                let env_value = self
+                    .builder
+                    .build_struct_gep(
+                        function_ctx.env_type.unwrap(),
+                        env_ptr.into_pointer_value(),
+                        i as u32,
+                        "",
+                    )
+                    .unwrap();
+
+                v.replace(self.builder.build_load(candy_type_ptr, env_value, ""));
+            }
+        }
+        if v.is_none() {
+            if let Some(value) = self.locals.get(id) {
+                v.replace(*value.clone());
+            }
+        }
```

Comment:
With `#![feature(let_chains)]`:

```suggestion
        if v.is_none() && let Some(i) = function_ctx.captured_ids.iter().position(|i| i == id) {
            let env_ptr = function_ctx.function_value.get_last_param().unwrap();

            let env_value = self
                .builder
                .build_struct_gep(
                    function_ctx.env_type.unwrap(),
                    env_ptr.into_pointer_value(),
                    i as u32,
                    "",
                )
                .unwrap();

            v = Some(self.builder.build_load(candy_type_ptr, env_value, ""));
        }
        if v.is_none() && let Some(value) = self.locals.get(id) {
            v = Some(*value.clone());
        }
```